Source/WebCore/editing/cocoa/HTMLConverter.mm

```diff
@@ -2442,6 +2443,29 @@ AttributedString attributedString(const SimpleRange& range)
     return HTMLConverter { range }.convert();
 }
 
+#if ENABLE(WRITING_TOOLS)
+static bool hasBlockquoteOrPreAncestor(const Node& node)
+{
+    auto* element = node.parentElement();
+    while (element) {
+        auto hasBlockquoteOrPreTag = [&] {
+            RefPtr htmlElement = dynamicDowncast<HTMLElement>(element);
+            if (!htmlElement)
+                return false;
+
+            return htmlElement->hasTagName(blockquoteTag) || htmlElement->hasTagName(preTag);
+        }();
+
+        if (hasBlockquoteOrPreTag)
+            return true;
+
+        element = element->parentElement();
```

Comment:
I'm still a little concerned about this approach. We're going to be doing a lot of redundant DOM traversal.

Reply:
I added a caching mechanism